rootfs/etc/nginx/lua/balancer.lua

```diff
@@ -122,9 +122,54 @@ local function sync_backends()
   end
 end
 
+local function route_to_alternative_balancer(balancer)
+  if not balancer.alternative_backends then
+    return false
+  end
+
+  -- TODO: support traffic shaping for n > 1 alternative backends
+  local alternative_balancer = balancers[balancer.alternative_backends[1]]
+
+  local clean_target_header = util.replace_special_char(alternative_balancer.traffic_shaping_policy.header, "-", "_")
+
+  local header = ngx.var["http_" .. clean_target_header]
+  if header then
+    if header == "always" then
+      return true
+    elseif header == "never" then
+      return false
+    end
+  end
+
+  local clean_target_cookie = util.replace_special_char(alternative_balancer.traffic_shaping_policy.cookie, "-", "_")
+
+  local cookie = ngx.var["cookie_" .. clean_target_cookie]
+  if cookie then
+    if cookie == "always" then
+      return true
+    elseif cookie == "never" then
+      return false
+    end
+  end
+
+  if math.random(100) <= alternative_balancer.traffic_shaping_policy.weight then
+    return true
+  end
+
+  return false
+end
+
 local function get_balancer()
   local backend_name = ngx.var.proxy_upstream_name
-  return balancers[backend_name]
+
+  local balancer = balancers[backend_name]
+
```

Comment:
If `balancer` is `nil` we should short-circuit and return here. This is the case when all endpoints are unavailable or there's no endpoint configured.

Currently as you can see in the e2e tests, you will get an error like

```
018/11/02 07:56:39 [error] 228#228: *112 lua entry thread aborted: runtime error: /etc/nginx/lua/balancer.lua:126: attempt to index local 'balancer' (a nil value)
stack traceback:
coroutine 0:
	/etc/nginx/lua/balancer.lua: in function 'route_to_alternative_balancer'
	/etc/nginx/lua/balancer.lua:167: in function 'get_balancer'
	/etc/nginx/lua/balancer.lua:184: in function 'rewrite'
	rewrite_by_lua(nginx.conf:422):2: in function <rewrite_by_lua(nginx.conf:422):1>, client: 10.244.2.1, server: unavailable.svc.com, request: "GET / HTTP/1.1", host: "unavailable.svc.com"
2018/11/02 07:56:39 [error] 228#228: *112 failed to run log_by_lua*: /etc/nginx/lua/balancer.lua:126: attempt to index local 'balancer' (a nil value)
stack traceback:
	/etc/nginx/lua/balancer.lua:126: in function 'route_to_alternative_balancer'
	/etc/nginx/lua/balancer.lua:167: in function 'get_balancer'
	/etc/nginx/lua/balancer.lua:212: in function 'log'
```

Reply:
Fixed 👍